### DISCUSSION TITLE
URL parsed from a serialised request is never absolute

### DISCUSSION THREAD

**Discussion by choroba** [mojolicious/mojo#2142] [General]
* Mojolicious version: 9.35
* Perl version: 5.26.1
* Operating system: Linux OpenSUSE Leap 15.5 (5.14.21-150500.55.39-default)

### Steps to reproduce the behavior
```perl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use Mojo::Message::Request;
use Mojo::URL;

my $url1 = 'Mojo::URL'->new('http://example.com');
say $url1->to_string, ' ', $url1->is_abs ? 'abs' : '!abs';  # Is absolute.

my $txt = join "\r\n", 'GET / HTTP/1.1', 'Host: example.com', "", "";
my $url2 = 'Mojo::Message::Request'->new->parse($txt)->url;
say $url2->to_string, ' ', $url2->is_abs ? 'abs' : '!abs';  # Isn't absolute!
```
### Expected behavior
`$url2` should be absolute.

### Actual behavior
`$url2` is not absolute.

This leads to unexpected behaviour: when a serialised (`to_string`) request is parsed into a request and this request is `start`ed, it fails.
See [this StackOverflow question](https://stackoverflow.com/q/77748695/1030675).

**Comment by kraih** [comment]
I'm afraid your assumption is incorrect, the URL is not meant to be absolute. Mojolicious web apps are assumed to be deployed relative to a prefix for historic reasons going back to CGI. So you need to use `to_abs`.

**Reply by choroba** [reply]
Do you mean 
```perl
    $tx->req->url($tx->req->url->to_abs);
```
instead of my "hack" in the StackOverflow question?

**Comment by jberger** [comment]
To expand on what kraih said slightly (and because I didn't see his reply before trying it out 😄) the absolute parts of the url are contained within the url's base, and therefore you need to use to_abs.

```bash
$ perl -Mojo -E 'say Mojo::Message::Request->new->parse(join "\r\n", "GET / HTTP/1.1", "Host: example.com", "", "")->url->to_abs'
http://example.com/
```

**Reply by choroba** [reply]
Do you think my solution to the SO question is not a hack, then?

**Reply by jberger** [reply]
Not if you want to store it in the request url as absolute, no. As kraih said, though, you might not actually want to store it there in that way if you're doing things like mounting applications at a certain path etc.
